### PR TITLE
Fixed #29789 -- Added support for nested relations to FilteredRelation.

### PR DESCRIPTION
**Pull request by ferrants**
For ticket 29789
https://code.djangoproject.com/ticket/29789

### CODE REVIEW

File: tests/filtered_relation/tests.py [hunk 1]
```diff
@@ -156,8 +156,15 @@ def test_with_multiple_filter(self):
...
-            ).filter(book_title_alice__isnull=False).filter(book_title_alice__isnull=False).distinct(),
+                book_title_alice=FilteredRelation(
+                    'book',
+                    condition=Q(book__title__icontains='alice')
+                ),
+            ).filter(
+                book_title_alice__isnull=False
+            ).filter(
+                book_title_alice__isnull=False
+            ).distinct(),
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: This change is unrelated, please revert it.

File: tests/filtered_relation/tests.py [hunk 2]
```diff
@@ -248,10 +255,11 @@ def test_difference(self):
...
 
     def test_select_for_update(self):
+        qs = Author.objects.annotate(
+            book_jane=FilteredRelation('book', condition=Q(book__title__iexact='the book by jane a')),
+        ).filter(book_jane__isnull=False).select_for_update()
         self.assertSequenceEqual(
-            Author.objects.annotate(
-                book_jane=FilteredRelation('book', condition=Q(book__title__iexact='the book by jane a')),
-            ).filter(book_jane__isnull=False).select_for_update(),
+            qs,
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: This change is unrelated, please revert it.

File: tests/filtered_relation/tests.py [hunk 3]
```diff
@@ -279,28 +287,153 @@ def test_as_subquery(self):
...
+            },
+        ])
+
+    def test_with_composed_relations(self):
+        qs = Author.objects.annotate(
+            my_books=FilteredRelation(
+                'book', condition=Q(book__title__icontains='book')
+            )
+        )
+        qs = qs.annotate(
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: I would prefer to chain `annotate()`, e.g.
```python
...
).annotate(
...
```

File: django/db/models/sql/query.py [hunk 4]
```diff
@@ -1408,13 +1408,23 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
...
+            )
+        for lookup in chain(lookups):
             lookup_parts, field_parts, _ = self.solve_lookup_type(lookup)
             shift = 2 if not lookup_parts else 1
-            if len(field_parts) > (shift + len(lookup_parts)):
+            lookup_path = field_parts[:-shift]
+            for _ in relation_field_parts:
+                if lookup_path:
+                    lookup_path = lookup_path[1:]
+            if lookup_path:
```
Comment (felixxm) [inline django/db/models/sql/query.py]: The above five lines can be simplified to:
```python
    ...
    shift = 2 if not lookup_parts else 1
    if len(field_parts) > (shift + len(relation_field_parts)):
    raise ValueError(...
```
so we don't need a `lookup_path` variable and iterate over the `relation_field_parts`.

File: tests/filtered_relation/tests.py (hunk 2 above)
Comment (ferrants) [reply tests/filtered_relation/tests.py]: addressed

File: tests/filtered_relation/tests.py (hunk 3 above)
Comment (ferrants) [reply tests/filtered_relation/tests.py]: addressed!

File: django/db/models/sql/query.py (hunk 4 above)
Comment (ferrants) [reply django/db/models/sql/query.py]: addressed!

**Comment by ferrants** [comment]
@felixxm , i addressed your feedback and have rebased. any other feedback is greatly appreciated!

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -1408,13 +1408,19 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
...
+            )
+        for lookup in chain(lookups):
             lookup_parts, field_parts, _ = self.solve_lookup_type(lookup)
             shift = 2 if not lookup_parts else 1
-            if len(field_parts) > (shift + len(lookup_parts)):
+            if len(field_parts) > (shift + len(relation_field_parts)):
                 raise ValueError(
                     "FilteredRelation's condition doesn't support nested "
-                    "relations (got %r)." % lookup
+                    "on clauses deeper than the relation (got %r for %r)." % (lookup, filtered_relation.relation_name)
```
Comment (felixxm) [inline django/db/models/sql/query.py]: Validation based only on a nesting level is not a proper solution, e.g.
```python
Author.objects.annotate(
    book_editor_worked_with=FilteredRelation('book__editor', condition=Q(book__author__name__icontains='book')),
)
``` 
will crash with `django.db.utils.OperationalError: no such column: T4.name`.

File: tests/filtered_relation/tests.py [hunk 6]
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
...
         with self.assertRaisesMessage(ValueError, msg):
-            list(Author.objects.annotate(
-                book_edited_by_b=FilteredRelation('book', condition=Q(book__editor__name__icontains='b')),
-            ))
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation(
+                    'book__title__icontains',
+                    condition=Q(book__editor__name__icontains='b')
+                ),
+            ).filter(book_edited_by_b__isnull=False).distinct()
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: `distinct()` is not necessary for this test, please remove it.

File: tests/filtered_relation/tests.py [hunk 7]
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
...
         with self.assertRaisesMessage(ValueError, msg):
-            list(Book.objects.annotate(
-                alice_favorite_books=FilteredRelation(
-                    'author__favorite_books',
-                    condition=Q(author__favorite_books__author=self.author1),
-                )
-            ))
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation('book', condition=Q(book__editor__name__icontains='b')),
+            ).filter(book_edited_by_b__isnull=False).distinct()
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: `distinct()` is not necessary for this test, please remove it.

File: tests/filtered_relation/tests.py [hunk 8]
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
...
+
+        with self.assertNumQueries(1):
+            self.assertQuerysetEqual(qs, [
+                (self.author1, self.editor_a),
+                (self.author2, self.editor_b),
+            ], lambda x: (x, x.book_editor_worked_with))
+
+    def test_with_nested_field(self):
+        qs = Author.objects.annotate(
+            book_editor_worked_with=FilteredRelation('book__editor', condition=Q(book__title__icontains='book')),
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: I'm not sure if that is a correct example and sth that we would like to support :thinking:, because conditions across different relations introduce duplicate rows. IMO we should restrict conditions only to starting with `filtered_relation`.

Comment (ferrants) [reply tests/filtered_relation/tests.py]: What about restricting to those within the `filtered_relation` path? `book__title__icontains` comes from the `book` table, which is accessed on the way to `book_editor`

For more context, I'm using this to do a join on the right side of a left join to convert currencies. I could get more into my specific models if that'd be helpful, but don't want them to distract from the basic functionality. I didn't want to add to the models under test here to replicate what I'm doing, but maybe I should to make it more explicit.

File: django/db/models/sql/query.py [hunk 9]
```diff
@@ -1408,13 +1408,41 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
...
+                "FilteredRelation's relation_name doesn't support lookups "
+                "(got %r)." % filtered_relation.relation_name
+            )
+        for lookup in chain(lookups):
             lookup_parts, field_parts, _ = self.solve_lookup_type(lookup)
             shift = 2 if not lookup_parts else 1
-            if len(field_parts) > (shift + len(lookup_parts)):
+            field_path = field_parts[:-shift]
+
+            if len(field_path) <= len(relation_field_parts):
```
Comment (ferrants) [inline django/db/models/sql/query.py]: I reverted this back to looping through the relationship and lookup fields so I can have better checks to make sure the lookup is in the relationship path. There is probably more I should add to this. Thinking about it.

File: django/db/models/sql/query.py [hunk 10]
```diff
@@ -1408,13 +1408,41 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
...
+                    elif not nested_field_path:
+                        nested_field_path = True
+                    else:
+                        raise ValueError(
+                            "FilteredRelation's condition doesn't support nested "
+                            "on clauses deeper than the relation (got %r for %r)." % (
+                            lookup, filtered_relation.relation_name)
+                        )
+            else:
                 raise ValueError(
```
Comment (ferrants) [inline django/db/models/sql/query.py]: Probably not the right error case, it's also duplicate of the one right above it, checking inside the nested one. This can probably be removed, but I am considering if there are other errors or checks that should be made.

File: django/db/models/sql/query.py (hunk 5 above)
Comment (ferrants) [reply django/db/models/sql/query.py]: I'll add a test case specifically for this, thanks

Comment (ferrants) [reply django/db/models/sql/query.py]: Added test to catch this error

File: tests/filtered_relation/tests.py (hunk 8 above)
Comment (ferrants) [reply tests/filtered_relation/tests.py]: @felixxm , I added restrictions around this so the `filtered_relation`'s condition must be within the path to the `filtered_relation`

**Comment by ferrants** [comment]
@felixxm , rebased and updated the FilteredRelation condition check to be more strict, enforcing it to be within the path to the FilteredRelation, along with some more tests. Could you take another look? Thanks!

**Comment by ferrants** [comment]
@felixxm , I've rebased this, can you take another look? Thanks!

**Comment by ferrants** [comment]
@felixxm , I've rebased this and updated the ticket, can someone take another look? Is there another way I should be reaching out about this? Thanks!

**Comment by auvipy** [comment]
you should assign the ticket to you comment there. I think it's already in the review Q

**Comment by ferrants** [comment]
@auvipy , good call, thanks. Assigned to me

**Comment by ferrants** [comment]
@felixxm , I've rebased this again, any chance I could get another review? Thanks!

File: django/db/models/sql/query.py [hunk 11]
```diff
@@ -1407,14 +1407,32 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
...
-                raise ValueError(
-                    "FilteredRelation's condition doesn't support nested "
-                    "relations (got %r)." % lookup
-                )
+            lookup_field_path = lookup_field_parts[:-shift]
+            for idx, lookup_field_part in enumerate(lookup_field_path):
+                if len(relation_field_parts) > idx:
+                    if relation_field_parts[idx] != lookup_field_part:
+                        raise ValueError(
+                            "FilteredRelation's condition doesn't support nested "
```
[the remark this answers is not in the document]
Comment (ferrants) [reply django/db/models/sql/query.py]: Good call, thanks

**Comment by ferrants** [comment]
Rebased! Still good for review and merge!

**Comment by ferrants** [comment]
Rebased, still good to go! Been using this patch for months and it does the trick. Nested FilteredRelations are pretty useful for analytical queries.

File: tests/filtered_relation/tests.py [hunk 12]
```diff
@@ -279,28 +279,184 @@ def test_as_subquery(self):
...
+            "(got 'book__title__icontains')."
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation(
+                    'book__title__icontains',
+                    condition=Q(book__editor__name__icontains='b')
+                ),
+            ).filter(book_edited_by_b__isnull=False)
+            list(qs)
```
Comment (adamchainz) [inline tests/filtered_relation/tests.py:358]: Afaics this message should be triggered in the `.annotate()`, so the `filter()` and `list()` aren't needed?

File: tests/filtered_relation/tests.py [hunk 13]
```diff
@@ -279,28 +279,184 @@ def test_as_subquery(self):
...
-            list(Book.objects.annotate(
-                alice_favorite_books=FilteredRelation(
-                    'author__favorite_books',
-                    condition=Q(author__favorite_books__author=self.author1),
-                )
-            ))
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation('book', condition=Q(book__editor__name__icontains='b')),
+            ).filter(book_edited_by_b__isnull=False)
+            list(qs)
```
Comment (adamchainz) [inline tests/filtered_relation/tests.py:344]: Afaics this message should be triggered in the `.annotate()`, so the `filter()` and `list()` aren't needed?

**Comment by ferrants** [comment]
Thanks @adamchainz ! I addressed your comments and added to the 3.2 release notes

File: tests/filtered_relation/tests.py [hunk 14]
```diff
@@ -279,28 +286,182 @@ def test_as_subquery(self):
...
+        )
+        with self.assertRaisesMessage(ValueError, msg):
+            Author.objects.annotate(
+                book_edited_by_b=FilteredRelation(
+                    'book__title__icontains',
+                    condition=Q(book__editor__name__icontains='b')
+                ),
+            ).filter(book_edited_by_b__isnull=False)
+
+    def test_with_foreign_key_relation_name_lookup_not_within_path(self):
```
Comment (ferrants) [inline tests/filtered_relation/tests.py:365]: @charettes , this test asserts that the "right side" of the condition can't be deeper, which should avoid to N-to-Many relationship nested query issue.

File: tests/filtered_relation/tests.py [hunk 15]
```diff
@@ -279,28 +286,182 @@ def test_as_subquery(self):
...
+            'pk', 'author_favorite_book_editor__pk'
+        )
+
+        with self.assertNumQueries(1):
+            self.assertQuerysetEqual(qs, [
+                (self.book1, self.editor_b),
+                (self.book4, self.editor_b),
+            ], lambda x: (x, x.author_favorite_book_editor))
+
+    def test_with_foreign_key_on_condition_deeper_than_relationship_error(self):
```
Comment (ferrants) [inline tests/filtered_relation/tests.py:342]: @charettes , I've considered your comment and I think that the current limitations around lookup usages prevent that from happening. This test asserts that the "left side" of the condition can't be deeper, which should avoid to N-to-Many relationship nested query issue.

File: tests/filtered_relation/tests.py [hunk 16]
```diff
@@ -424,3 +585,182 @@ def test_aggregate(self):
...
+        )
+        BookDailySales.objects.create(
+            book=cls.book2,
+            date=cls.sales_date2,
+            currency_code=cls.currency2,
+            sales=100.00,
+            seller=cls.seller2,
+        )
+
+    def test_with_summation_aggregation(self):
```
Comment (ferrants) [inline tests/filtered_relation/tests.py:700]: This test reflects how I am actually using this feature to get more performance out of some queries by improving the join-on logic.

File: tests/filtered_relation/tests.py [hunk 17]
```diff
@@ -424,3 +585,182 @@ def test_aggregate(self):
...
+            {
+                'title': self.book3.title, 'sales_sum': None,
+            },
+        ])
+
+    def test_with_summation_aggregation_and_negation(self):
+        qs = Book.objects.annotate(
+            recent_sales=FilteredRelation(
+                'daily_sales',
+                condition=~Q(daily_sales__seller=self.seller1)
```
Comment (ferrants) [inline tests/filtered_relation/tests.py:738]: @charettes , this case is most similar to what you suggested we have tests cover, using negation on a nested relationship.

Comment (charettes) [reply tests/filtered_relation/tests.py:738]: This code path won't trigger `split_exclude` but from my local testing it seems `condition=~Q` already has a flawed implementation on `master` anyway.

```diff
     def test_with_exclude(self):
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=True),
+            [self.author2],
+        )
         self.assertSequenceEqual(
             Author.objects.annotate(
                 book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
             ).exclude(book_alice__isnull=False),
             [self.author2],
         )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=False),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).exclude(book_alice__isnull=True),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).filter(book_not_alice__isnull=False),
+            [self.author2, self.author2, self.author1],
+        )
+        # The following fails while all the other succeed.
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).exclude(book_not_alice__isnull=True),
+            [self.author2, self.author2, self.author1],
+        )

     def test_with_join_and_complex_condition(self):
         self.assertSequenceEqual(
```

Comment (charettes) [reply tests/filtered_relation/tests.py:738]: To summarize the issue mentioned above is not related to the patch here, looks like it already exists in master.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/filtered_relation/tests.py (hunk 17 above)
Comment (ferrants) [reply tests/filtered_relation/tests.py:738]: Thanks for the `--debug-sql` tip, I wasn't aware of that. And thanks for those examples of what you mean. It looks like there was a failure in the `database=oracle18,label=bionic4,python=python3.8` tests and now after rebasing in the spelling checks in the docs, so I'll look into those tests today as well to see if I can get a better assertion in place.

Comment (ferrants) [reply tests/filtered_relation/tests.py:738]: @charettes , Some other tests in this suite use a `distinct` to I think avoid the same issue, would you feel comfortable using a distinct to assert the right result? For example, in the last example that you posted, if you add a distinct, this passes:
```
        self.assertSequenceEqual(
            Author.objects.annotate(
                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
            ).filter(book_not_alice__isnull=False).order_by('id').distinct(),
            [self.author1, self.author2],
        )
```
with query:
```
SELECT DISTINCT "filtered_relation_author"."id", "filtered_relation_author"."name", "filtered_relation_author"."content_type_id", "filtered_relation_author"."object_id" FROM "filtered_relation_author" INNER JOIN "filtered_relation_book" book_not_alice ON ("filtered_relation_author"."id" = book_not_alice."author_id" AND (NOT (UPPER(book_not_alice."title"::text) = UPPER(poem by alice)))) WHERE book_not_alice."id" IS NOT NULL ORDER BY "filtered_relation_author"."id" ASC
```

I don't think it make sense to assert the `[self.author1, self.author2, self.author2]` result for the non-distinct one just to document the current behavior, we should probably try to push-down to a sub-query in that case, right? Do we have a separate ticket for the existing issue in `master`? There are some closed-fixed ones here that look related: https://code.djangoproject.com/search?q=split_exclude&noquickjump=1&changeset=on&milestone=on&ticket=on&wiki=on

Comment (charettes) [reply tests/filtered_relation/tests.py:738]: @ferrants I think this can be addressed in a separate ticket yes, the above diff was more a of demonstration of current issues with negated filtered conditions but you're probably right that it's related to existing tickets.

**Comment by felixxm** [comment]
@ferrants I cannot push my final edits :disappointed: Can you check 
obraz?

**Comment by felixxm** [comment]
I pushed final edits to a new PR #13170
